lib/taskana-core/src/main/java/pro/taskana/jobs/WorkbasketCleanupJob.java

```diff
@@ -0,0 +1,172 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import pro.taskana.BaseQuery;
+import pro.taskana.BulkOperationResults;
+import pro.taskana.TaskService;
+import pro.taskana.TaskState;
+import pro.taskana.TaskanaEngine;
+import pro.taskana.exceptions.InvalidArgumentException;
+import pro.taskana.exceptions.NotAuthorizedException;
+import pro.taskana.exceptions.TaskanaException;
+import pro.taskana.exceptions.WorkbasketInUseException;
+import pro.taskana.exceptions.WorkbasketNotFoundException;
+import pro.taskana.impl.util.LoggerUtils;
+import pro.taskana.transaction.TaskanaTransactionProvider;
+
+/**
+ * Job to cleanup completed workbaskets after a period of time if there are no pending tasks associated to the workbasket.
+ */
+public class WorkbasketCleanupJob extends AbstractTaskanaJob {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TaskCleanupJob.class);
+
+    // Parameter
+    private Instant firstRun;
+    private Duration runEvery;
+    private int batchSize;
+
+    public WorkbasketCleanupJob(TaskanaEngine taskanaEngine,
+        TaskanaTransactionProvider<Object> txProvider, ScheduledJob job) {
+        super(taskanaEngine, txProvider, job);
+        firstRun = taskanaEngine.getConfiguration().getCleanupJobFirstRun();
+        runEvery = taskanaEngine.getConfiguration().getCleanupJobRunEvery();
+        batchSize = taskanaEngine.getConfiguration().getMaxNumberOfUpdatesPerTransaction();
+    }
+
+    @Override
+    public void run() throws TaskanaException {
+        LOGGER.info("Running job to delete all workbaskets marked for deletion");
+        try {
+            List<String> workbasketsMarkedForDeletion = getWorkbasketsMarkedForDeletion();
+            int totalNumberOfWorkbasketDeleted = 0;
+            while (workbasketsMarkedForDeletion.size() > 0) {
+                int upperLimit = batchSize;
+                if (upperLimit > workbasketsMarkedForDeletion.size()) {
+                    upperLimit = workbasketsMarkedForDeletion.size();
+                }
+                totalNumberOfWorkbasketDeleted += deleteWorkbasketsTransactionally(
+                    workbasketsMarkedForDeletion.subList(0, upperLimit));
+                workbasketsMarkedForDeletion.subList(0, upperLimit).clear();
+            }
+            LOGGER.info("Job ended successfully. {} workbaskets deleted.", totalNumberOfWorkbasketDeleted);
+        } catch (Exception e) {
+            throw new TaskanaException("Error while processing WorkbasketCleanupJob.", e);
+        } finally {
+            scheduleNextCleanupJob();
+        }
+    }
+
+    private List<String> getWorkbasketsMarkedForDeletion() throws InvalidArgumentException {
+        List<String> workbasketList = taskanaEngineImpl.getWorkbasketService()
+            .createWorkbasketQuery()
+            .markedForDeletion(true)
+            .listValues("ID", BaseQuery.SortDirection.ASCENDING);
+        workbasketList = excludeWorkbasketWithPendingTasks(workbasketList);
```

Comment:
This is a good idea. But wouldn't this be better placed in the deleteWorkbaskets() operation instead of the cleanup job?